Cleanup deprecated methods in wait package

**What would you like to be added**:

- [x] Update `wait.Poll` to `wait.PollUntilContextTimeout` (with `immediately = false`)  (@Affan-7, #3906 )
For example:
```diff
--- a/pkg/controllers/certificate/cert_rotation_controller.go
+++ b/pkg/controllers/certificate/cert_rotation_controller.go
@@ -165,7 +165,7 @@ func (c *CertRotationController) syncCertRotation(secret *corev1.Secret) error {

        var newCertData []byte
        klog.V(1).Infof("Waiting for the client certificate to be issued")
-       err = wait.Poll(1*time.Second, 5*time.Minute, func() (done bool, err error) {
+       err = wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, 5*time.Minute, false, func(ctx context.Context) (done bool, err error) {
                csr, err := c.KubeClient.CertificatesV1().CertificateSigningRequests().Get(context.TODO(), csr, metav1.GetOptions{})
                if err != nil {
                        return false, fmt.Errorf("failed to get the cluster csr %s. err: %v", clusterName, err)
```

- [x] Update `wait.PollImmediate` to `wait.PollUntilContextTimeout` (with `immediately = true`)  (@liangyuanpeng, #3921)
For example:
```diff
diff --git a/pkg/controllers/cluster/cluster_controller.go b/pkg/controllers/cluster/cluster_controller.go
index ae123a86e..678bf7188 100644
--- a/pkg/controllers/cluster/cluster_controller.go
+++ b/pkg/controllers/cluster/cluster_controller.go
@@ -394,7 +394,7 @@ func (c *Controller) monitorClusterHealth(ctx context.Context) (err error) {
        for i := range clusters {
                cluster := &clusters[i]
                var observedReadyCondition, currentReadyCondition *metav1.Condition
-               if err = wait.PollImmediate(MonitorRetrySleepTime, MonitorRetrySleepTime*HealthUpdateRetry, func() (bool, error) {
+               if err = wait.PollUntilContextTimeout(ctx, MonitorRetrySleepTime, MonitorRetrySleepTime*HealthUpdateRetry, true, func(ctx context.Context) (bool, error) {
                        // Cluster object may be changed in this function.
                        observedReadyCondition, currentReadyCondition, err = c.tryUpdateClusterHealth(ctx, cluster)
                        if err == nil {

```

- [x] Update `wait.PollUntil` to `PollUntilContextCancel` (@zhzhuang-zju, [#4947](https://github.com/karmada-io/karmada/pull/4947))

This might be a little bit challenging, we need to transition channel to context. Probably could leverage [ContextForChannel](https://github.com/karmada-io/karmada/blob/e5277b6317ac1a4717f5fac4057caf51a5d248fc/pkg/util/context.go#L11).

- [x] Enable static check by removing [these lines](https://github.com/karmada-io/karmada/blame/e5277b6317ac1a4717f5fac4057caf51a5d248fc/.golangci.yml#L69-L72) (@RainbowMango, #4992)


**Why is this needed**:
The `wait.PollImmediate`, `wait.Poll`, etc methods were deprecated in Kubernetes v1.27 ([#107826](https://github.com/kubernetes/kubernetes/pull/107826)), lint check is falling during update the dependencies of Kubernetes at #3730.
This kind of check has been disabled by #3730 to avoid huge modifications.

```
test/e2e/framework/cluster.go:320:9: SA1019: wait.PollImmediate is deprecated: This method does not return errors from context, use PollWithContextTimeout. Note that the new method will no longer return ErrWaitTimeout and instead return errors defined by the context package. Will be removed in a future release. (staticcheck)
	return wait.PollImmediate(2*time.Second, 10*time.Second, func() (done bool, err error) {
```

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -82,10 +82,6 @@ linters-settings:
       - name: unexported-return
       - name: time-naming
       - name: empty-block
-  staticcheck:
-    checks:
-      - all
-      - '-SA1019' # disable deprecation check. Tracked by https://github.com/karmada-io/karmada/issues/3835.
 
 issues:
   # The list of ids of default excludes to include or disable. By default it's empty.

--- a/hack/verify-staticcheck.sh
+++ b/hack/verify-staticcheck.sh
@@ -19,7 +19,7 @@ set -o nounset
 set -o pipefail
 
 REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
-GOLANGCI_LINT_VER="v1.58.0"
+GOLANGCI_LINT_VER="v1.59.0"
 
 cd "${REPO_ROOT}"
 source "hack/util.sh"

